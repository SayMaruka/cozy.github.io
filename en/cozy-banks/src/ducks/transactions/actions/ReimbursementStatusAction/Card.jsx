import React from 'react'
import PropTypes from 'prop-types'
import { translate } from 'cozy-ui/react'
import UICard from 'cozy-ui/react/Card'
import { Caption, Text } from 'cozy-ui/react/Text'
import Icon, { iconPropType } from 'cozy-ui/react/Icon'
import cx from 'classnames'
import styles from 'ducks/transactions/actions/ReimbursementStatusAction/Card.styl'
import { getPlatform } from 'cozy-device-helper'

// TODO use icons from cozy-ui when https://github.com/cozy/cozy-ui/pull/983
// has been merged
import iconPhone from 'assets/icons/icon-phone.svg'
import iconEmail from 'assets/icons/icon-email.svg'

const Card = props => {
  const { title, caption, icon, className, ...rest } = props

  return (
    <UICard className={cx(styles.Card, className)} {...rest}>
      <Icon icon={icon} size={16} color="var(--coolGrey)" />
      <div className="u-ml-1">
        <Text>{title}</Text>
        <Caption>{caption}</Caption>
      </div>
    </UICard>
  )
}

Card.propTypes = {
  icon: iconPropType.isRequired,
  title: PropTypes.string.isRequired,
  caption: PropTypes.node.isRequired
}

export const DumbPhoneCard = props => {
  const { contact, t, ...rest } = props
  const href = `tel:${contact.number}`
  const caption = contact.price ? (
    <>
      {contact.price}
      {t('ReimbursementStatusModal.contact.phoneCallPrice')}
    </>
  ) : (
    t('ReimbursementStatusModal.contact.phoneCallFree')
  )

  return (
    <Card
      href={href}
      title={contact.number}
      caption={caption}
      icon={iconPhone}
      tag="a"
      {...rest}
    />
  )
}

const phoneContactShape = PropTypes.shape({
  price: PropTypes.string,
  number: PropTypes.string.isRequired
})

DumbPhoneCard.propTypes = {
  contact: phoneContactShape.isRequired,
  t: PropTypes.func.isRequired
}

export const PhoneCard = translate()(DumbPhoneCard)

export const DumbWebCard = props => {
  const { contact, t, ...rest } = props
  const caption = new URL(contact.href).hostname

  return (
    <Card
      href={contact.href}
      title={t(`ReimbursementStatusModal.contact.actions.${contact.action}`)}
      caption={caption}
      icon={iconEmail}
      target="_blank"
      tag="a"
      {...rest}
    />
  )
}

const webContactShape = PropTypes.shape({
  href: PropTypes.string.isRequired,
  action: PropTypes.string.isRequired
})

DumbWebCard.propTypes = {
  contact: webContactShape.isRequired,
  t: PropTypes.func.isRequired
}

export const WebCard = translate()(DumbWebCard)

export const DumbAppCard = props => {
  const { contact, t, ...rest } = props

  if (getPlatform() !== contact.platform) {
    return null
  }

  return (
    <Card
      href={contact.href}
      title={t('ReimbursementStatusModal.contact.openApp.title')}
      caption={t('ReimbursementStatusModal.contact.openApp.caption')}
      icon="phone"
      target="_blank"
      tag="a"
      {...rest}
    />
  )
}

const appContactShape = PropTypes.shape({
  platform: PropTypes.oneOf(['android', 'ios']).isRequired,
  href: PropTypes.string.isRequired
})

DumbAppCard.propTypes = {
  contact: appContactShape.isRequired,
  t: PropTypes.func.isRequired
}

export const AppCard = translate()(DumbAppCard)

export const ContactCard = props => {
  const { type, ...rest } = props

  switch (type) {
    case 'phone':
      return <PhoneCard {...rest} />

    case 'web':
      return <WebCard {...rest} />

    case 'app':
      return <AppCard {...rest} />

    default:
      return null
  }
}

ContactCard.propTypes = {
  type: PropTypes.oneOf(['phone', 'web', 'app']).isRequired,
  contact: PropTypes.object.isRequired
}
