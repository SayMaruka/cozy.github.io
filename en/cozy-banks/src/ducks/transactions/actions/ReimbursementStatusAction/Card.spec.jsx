import React from 'react'
import { DumbPhoneCard, DumbWebCard, DumbAppCard } from './Card'
import { mount } from 'enzyme'
import { getPlatform } from 'cozy-device-helper'

jest.mock('cozy-device-helper')

const t = key => key

describe('DumbPhoneCard', () => {
  it('should render correctly with number and price', () => {
    const contact = {
      number: '01 23 45 67 89',
      price: '0,06â‚¬/min'
    }

    expect(
      mount(<DumbPhoneCard contact={contact} t={t} />).html()
    ).toMatchSnapshot()
  })

  it('should render correctly with only number', () => {
    const contact = {
      number: '01 23 45 67 89'
    }

    expect(
      mount(<DumbPhoneCard contact={contact} t={t} />).html()
    ).toMatchSnapshot()
  })
})

describe('DumbWebCard', () => {
  it('should render correctly', () => {
    const contact = {
      href: 'https://ameli.fr',
      action: 'sendCareSheet'
    }

    expect(
      mount(<DumbWebCard contact={contact} t={t} />).html()
    ).toMatchSnapshot()
  })
})

describe('DumbAppCard', () => {
  it('should render nothing if the platform does not match with the current device', () => {
    getPlatform.mockReturnValueOnce('ios')

    const contact = { platform: 'android', href: 'http://some.thing' }

    expect(mount(<DumbAppCard contact={contact} t={t} />).html()).toBe(null)
  })

  it('should render correctly if the platform matches with the current device', () => {
    getPlatform.mockReturnValueOnce('ios')

    const contact = { platform: 'ios', href: 'http://some.thing' }
    expect(
      mount(<DumbAppCard contact={contact} t={t} />).html()
    ).toMatchSnapshot()
  })
})
