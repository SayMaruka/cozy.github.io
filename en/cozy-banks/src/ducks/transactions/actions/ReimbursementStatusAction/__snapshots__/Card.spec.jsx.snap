// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbAppCard should render correctly if the platform matches with the current device 1`] = `"<a class=\\"styles__c-card___YgP7B Card\\" href=\\"http://some.thing\\" target=\\"_blank\\"><svg class=\\"styles__icon___23x3R\\" style=\\"fill: var(--coolGrey);\\" width=\\"16\\" height=\\"16\\"><use xlink:href=\\"#phone\\"></use></svg><div class=\\"u-ml-1\\"><div class=\\"u-text\\">Open the application</div><div class=\\"u-caption\\">On my phone</div></div></a>"`;

exports[`DumbPhoneCard should render correctly with number and price 1`] = `"<a class=\\"styles__c-card___YgP7B Card\\" href=\\"tel:01 23 45 67 89\\"><svg class=\\"styles__icon___23x3R\\" style=\\"fill: var(--coolGrey);\\" width=\\"16\\" height=\\"16\\"><use xlink:href=\\"#test-file-stub\\"></use></svg><div class=\\"u-ml-1\\"><div class=\\"u-text\\">01 23 45 67 89</div><div class=\\"u-caption\\">0,06â‚¬/min + call price</div></div></a>"`;

exports[`DumbPhoneCard should render correctly with only number 1`] = `"<a class=\\"styles__c-card___YgP7B Card\\" href=\\"tel:01 23 45 67 89\\"><svg class=\\"styles__icon___23x3R\\" style=\\"fill: var(--coolGrey);\\" width=\\"16\\" height=\\"16\\"><use xlink:href=\\"#test-file-stub\\"></use></svg><div class=\\"u-ml-1\\"><div class=\\"u-text\\">01 23 45 67 89</div><div class=\\"u-caption\\">Not surcharged call</div></div></a>"`;

exports[`DumbWebCard should render correctly 1`] = `"<a class=\\"styles__c-card___YgP7B Card\\" href=\\"https://ameli.fr\\" target=\\"_blank\\"><svg class=\\"styles__icon___23x3R\\" style=\\"fill: var(--coolGrey);\\" width=\\"16\\" height=\\"16\\"><use xlink:href=\\"#test-file-stub\\"></use></svg><div class=\\"u-ml-1\\"><div class=\\"u-text\\">Send my care sheet</div><div class=\\"u-caption\\">ameli.fr</div></div></a>"`;
